File: level_nodes.cpp
//{ Driver Code Starts
#include<bits/stdc++.h>
using namespace std;

// } Driver Code Ends

class Solution
{
	public:
	//Function to find the level of node X.
	int nodeLevel(int V, vector<int> adj[], int X) 
	{
	    // code here
	    if (X >= V) return -1;
	    vector<int> v(V);
	    v[0] = 1;
	    
	    vector<int> level(V, -1);
	    queue<int> Q;
	    Q.push(0);
	    Q.push(0);
	    while(!Q.empty()) {
	        int c = Q.front();
	        Q.pop();
	        int l = Q.front();
	        Q.pop();
	        for(auto w : adj[c]) {
	            if (!v[w]) {
	                v[w] = 1;
	                level[w] = l + 1;
	                Q.push(w);
	                Q.push(l+1);
	            }
	        }
	    }
	    return level[X];
	}
};

//{ Driver Code Starts.


int main()
{
    
    int t;
    cin >> t;
    while(t--)
    {
    	int V, E, X;
    	cin >> V >> E;

    	vector<int> adj[V];

    	for(int i = 0; i < E; i++)
    	{
    		int u, v;
    		cin >> u >> v;
    		adj[u].push_back(v);
    		adj[v].push_back(u);
    	}
    	cin>>X;

    	Solution obj;
    	cout << obj.nodeLevel(V, adj, X) << "\n";
    }

    return 0;
}


// } Driver Code Ends

File: graph/level_nodes.py
# User function Template for python3
from collections import deque


class Solution:
    # Function to find the level of node X.
    def nodeLevel(self, V, adj, X):
        # code here
        q = deque()
        level = 0
        q.append(0)
        visited = [0] * V
        visited[0] = 1

        while q:
            n = len(q)
            for _ in range(n):
                v = q.popleft()
                if v == X:
                    return level
                for w in adj[v]:
                    if visited[w] == 0:
                        q.append(w)
                        visited[w] = 1
            level += 1

        return -1


# {
# Driver Code Starts
# Initial Template for Python 3

if __name__ == "__main__":
    t = int(input())
    for i in range(t):
        V, E = list(map(int, input().strip().split()))
        adj = [[] for i in range(V)]
        for i in range(E):
            a, b = map(int, input().strip().split())
            adj[a].append(b)
            adj[b].append(a)
        X = int(input())
        ob = Solution()

        print(ob.nodeLevel(V, adj, X))
# } Driver Code Ends
